pycc/local.py

```diff
@@ -366,6 +360,277 @@ def _build_LPNO(self):
+                
+                # S_ij = lmo by pno_o
+                # S = ao by ao 
+                S_ij = self.C.T @ S @ C_ij[ij]                  
+
+                # ERI_ij folds in the overlap terms and transformed into semicanonical pno basis  
+
+                # ERI_ij = pno_o by pno_o 
+                ERI_ij.append(S_ij.T @ self.H.ERI[i,j] @ S_ij)
+
+                # Construction of the t2 amplitude using the transformed ERI and semicanonicalized orbital energies obtained from the 
+                # diagonalizing the pno fock matrix ... an alternative way is to obtain the initial t2 ampltiude and transform
+      
+                # t2_ij = pno_o by pno_o  
+                t2_ij.append( -1.0 * ERI_ij[ij] / (self.eps[ij].reshape(1,-1) + self.eps[ij].reshape(-1,1) - self.H.F[i,i] - self.H.F[j,j]))
+
+        #Note that the t_ij is specific to each pair so when you are doing the emp2 energy calculation
+        #either calculate it in the pair loop or revert back to the MO t2 and MO L
+   
+        # parameters for the MP2 loop
+        e_conv = 1e-7
+        r_conv = 1e-7
+        maxiter = 100
+                
+        emp2 = contract('ijab,ijab->', t2,self.H.L[o,o,v,v])
+        #print(emp2)
+        ediff = emp2
+        rmsd = 0.0
+        niter = 0
+
+        if local.upper() in ["LPNO"]:
+            while ((abs(ediff) > e_conv) or (abs(rmsd) > r_conv)) and (niter <= maxiter):
+                niter += 1
+                elast = emp2
+                emp2 = 0
+               
+                #zach's version
+                for ij in range(self.no*self.no):  
+                    i = ij // self.no 
+                    j = ij % self.no
+                    
+                    # self.eps[ij] contains the semicanonical orbital energies obtain through the diagonalization of the pno fock matrix                    
+                    r2_ij = ERI_ij[ij] + (self.eps[ij].reshape(-1,1) + self.eps[ij].reshape(1,-1) - self.H.F[i,i] - self.H.F[j,j])* t2_ij[ij]
+                    
+                    for k in range(self.no):
+                        if i != k: 
+                            kj = self.no * k + j 
+                            S_ijkj = C_ij[ij].T @ S @ C_ij[kj]
+                            r2_ij -= F_occ[i,k] * contract('rs,ra,sb->ab',t2_ij[kj],S_ijkj,S_ijkj)
+                        if j != k:
+                            ik = self.no * i + k 
+                            S_ijik = C_ij[ij].T @ S @ C_ij[ik]
+                            r2_ij -= F_occ[k,j] * contract('rs,ra,sb->ab',t2_ij[ik],S_ijik,S_ijik)
+
+                    t2_ij[ij] -= r2_ij / (self.eps[ij].reshape(-1,1) + self.eps[ij].reshape(1,-1) - self.H.F[i,i] - self.H.F[j,j])
+
+                    rmsd += np.sqrt(contract('ab,ab->', r2_ij, r2_ij))
+                
+                    t2[i,j] = Q[ij] @ L[ij] @ t2_ij[ij] @ L[ij].T @ Q[ij].T
+
+                emp2 = contract('ijab,ijab->', t2, self.H.L[o,o,v,v])
```

Comment:
it looks like the strategy here was to return to the MO basis amplitudes, so you can calculate the energy in one shot (instead of in a pair loop). however, I think that defeats the purpose of the production algorithm - you should never have to store or contract over something that is full-MO sized once you transform everything. the energy should be computed as a sum of pair energies.